File: tools/record_media_content.py
#!/usr/bin/env python3
"""Record the Python outputs in suites/opentelemetry-media-content/cases.json.

Runs prism-opentelemetry-py's real ``without_media_bytes()`` and writes the compact
JSON into the case file as TEXT, replacing only the ``"py"`` output literal.

    PRISM_OTEL_PY=../prism-opentelemetry-py python tools/record_media_content.py [--check]
"""

from __future__ import annotations

import json
import os
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
BRIDGE = Path(os.environ.get("PRISM_OTEL_PY", ROOT.parent / "prism-opentelemetry-py"))
sys.path.insert(0, str(BRIDGE / "src"))

from prism_opentelemetry import without_media_bytes  # noqa: E402

BACKSLASH = chr(92)


def literal_end(raw: str, start: int) -> int:
    """Where the JSON string literal whose opening quote is at ``start`` ends."""
    i = start + 1
    while i < len(raw):
        if raw[i] == BACKSLASH:
            i += 2
            continue
        if raw[i] == '"':
            return i + 1
        i += 1
    raise ValueError("Unterminated string literal.")


def main() -> int:
    check = "--check" in sys.argv
    path = ROOT / "suites" / "opentelemetry-media-content" / "cases.json"
    raw = path.read_text(encoding="utf-8")
    document = json.loads(raw)
    stale: list[str] = []

    for case in document["cases"]:
        output = json.dumps(
            without_media_bytes(case["input"]), separators=(",", ":"), ensure_ascii=False
        )
        if case["output"]["py"] != output:
            stale.append(case["id"])

        at = raw.index(f'"id": "{case["id"]}"')
        key = raw.index('"py": ', raw.index('"output": {', at))
        start = key + len('"py": ')
        raw = raw[:start] + json.dumps(output, ensure_ascii=False) + raw[literal_end(raw, start) :]

    if check:
        if stale:
            print(f"Stale py outputs: {', '.join(stale)}", file=sys.stderr)
            return 1
        print("PY outputs current.", file=sys.stderr)
        return 0

    path.write_text(raw, encoding="utf-8", newline="\n")
    print(f"Wrote {len(document['cases'])} py output(s); {len(stale)} changed.", file=sys.stderr)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/record-media-content.mjs
#!/usr/bin/env node
// Record the TypeScript outputs in suites/opentelemetry-media-content/cases.json,
// by running prism-opentelemetry-ts's real withoutMediaBytes(). Written into the
// case file as TEXT, replacing only the "ts" output literal.
//
//   PRISM_OTEL_TS=../prism-opentelemetry-ts node tools/record-media-content.mjs [--check]

import { readFileSync, writeFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';

const root = dirname(dirname(fileURLToPath(import.meta.url)));
const bridge = process.env.PRISM_OTEL_TS ?? join(root, '..', 'prism-opentelemetry-ts');
const { withoutMediaBytes } = await import(pathToFileURL(join(bridge, 'dist', 'index.js')).href);

/** Where the JSON string literal whose opening quote is at `open` ends, exclusive. */
function literalEnd(raw, open) {
  for (let i = open + 1; i < raw.length; i++) {
    if (raw[i] === '\\') {
      i++;
      continue;
    }
    if (raw[i] === '"') return i + 1;
  }
  throw new Error('Unterminated string literal.');
}

const check = process.argv.includes('--check');
const path = join(root, 'suites', 'opentelemetry-media-content', 'cases.json');
let raw = readFileSync(path, 'utf8');
const document = JSON.parse(raw);
const stale = [];

for (const testCase of document.cases) {
  const output = JSON.stringify(withoutMediaBytes(testCase.input));

  if (testCase.output.ts !== output) stale.push(testCase.id);

  const at = raw.indexOf(`"id": "${testCase.id}"`);
  const key = raw.indexOf('"ts": ', raw.indexOf('"output": {', at));
  const open = key + '"ts": '.length;
  raw = raw.slice(0, open) + JSON.stringify(output) + raw.slice(literalEnd(raw, open));
}

if (check) {
  if (stale.length > 0) {
    console.error(`Stale ts outputs: ${stale.join(', ')}`);
    process.exit(1);
  }
  console.error('TS outputs current.');
} else {
  writeFileSync(path, raw);
  console.error(`Wrote ${document.cases.length} ts output(s); ${stale.length} changed.`);
}
